cores/esp8266/TypeConversion.cpp

```diff
@@ -0,0 +1,82 @@
+/*
+   TypeConversion functionality
+   Copyright (C) 2019 Anders Löfgren
+
+   License (MIT license):
+
+   Permission is hereby granted, free of charge, to any person obtaining a copy
+   of this software and associated documentation files (the "Software"), to deal
+   in the Software without restriction, including without limitation the rights
+   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+   copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in
+   all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+   THE SOFTWARE.
+*/
+
+#include <assert.h>
+#include "TypeConversion.h"
+
+namespace esp8266 {
+  namespace TypeConversion {
+    const char base36Chars[36] = {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'};
+    const uint8_t base36CharValues[75] {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 0, 0, 0, 0, 0, 0, 0, // 0 to 10
+            10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 0, 0, 0, 0, 0, 0, // Upper case letters
+            10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35
+    }; // Lower case letters
+
+
+    String uint8ArrayToHexString(const uint8_t *uint8Array, const uint32_t arrayLength) {
+      String hexString;
+      if (!hexString.reserve(2 * arrayLength)) { // Each uint8_t will become two characters (00 to FF)
+        return emptyString;
+      }
+
+      for (uint32_t i = 0; i < arrayLength; ++i) {
+        hexString += base36Chars[ uint8Array[i] >> 4 ];
+        hexString += base36Chars[ uint8Array[i] % 16 ];
+      }
+
+      return hexString;
+    }
+
+    uint8_t *hexStringToUint8Array(const String &hexString, uint8_t *uint8Array, const uint32_t arrayLength) {
+      assert(hexString.length() >= arrayLength * 2); // Each array element can hold two hexString characters
+
+      for (uint32_t i = 0; i < arrayLength; ++i) {
+        uint8Array[i] = (base36CharValues[hexString.charAt(i * 2) - '0'] << 4) + base36CharValues[hexString.charAt(i * 2 + 1) - '0'];
+      }
+
+      return uint8Array;
+    }
+
+    uint8_t *uint64ToUint8Array(const uint64_t value, uint8_t *resultArray) {
+      resultArray[7] = value;
```

Comment:
I appreciate the clean C code, but on the 8266 couldn't this just reduce to `memcpy(resultArray, &value, 8)` or a simple pointer-cast and assign?

Reply:
A straight memory copy is not possible if we want to keep compatibility with how MAC addresses are handled in the Arduino Core at the moment (e.g. the conversion `MAC to uint64_t to String` would give a backwards result compared to `MAC to String`). This can be resolved by reversing the result of the memory copy, e.g. via `std::reverse`, but based on testing this would increase conversion time compared to the current approach by at least 75 % with the default compiler optimization flag (-Os).